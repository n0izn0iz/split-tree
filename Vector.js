import Space from "./Space";

export default class Vector extends Array {
  constructor(...values) {
    const space = values[0];
    if (space instanceof Space) super(...space.originVector);
    else {
      if (values.length === 1) {
        super(1);
        this[0] = values[0];
      } else super(...values);
    }
    this.__proto__ = Vector.prototype;
    if (this.length === 0) throw new Error("Created 0 dimensions Vector");
  }
  toString() {
    return `${this.length}DVector: { ${[...this]
      .map((value, index) => `${["x", "y", "z", "t"][index]}: ${value}`)
      .join(", ")} }`;
  }
  mapToVector(...args) {
    return new Vector(...this.map(...args));
  }
  plusScalar(scalar) {
    return this.mapToVector(value => value + scalar);
  }
  minusScalar(scalar) {
    return this.mapToVector(value => value - scalar);
  }
  plus(other) {
    if (Array.isArray(other)) other = new Vector(...other);
    const biggestVector = this.length > other.length ? this : other;
    const smallestVector = biggestVector === other ? this : other;
    return biggestVector.mapToVector(
      (value, index) =>
        value + (index < smallestVector.length ? smallestVector[index] : 0)
    );
  }
  scale(scalar) {
    return this.mapToVector(value => value * scalar);
  }
}
