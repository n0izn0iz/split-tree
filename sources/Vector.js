// @flow
export default class Vector extends Array<number> {
  constructor(...values: Array<number>) {
    if (values.length === 1) {
      super();
      this[0] = values[0];
    } else {
      super(...values);
    }
    // $FlowFixMe
    this.__proto__ = Vector.prototype;
    //if (this.length === 0) throw new Error("Created 0 dimensions Vector");
  }
  toString() {
    return `${this.length}DVector: { ${[...this]
      .map((value, index) => `${["x", "y", "z", "t"][index]}: ${value}`)
      .join(", ")} }`;
  }

  get array(): Array<number> {
    return [...this];
  }

  get time(): number {
    return this[this.length - 1];
  }

  get x(): number {
    return this[0];
  }
  get y(): number {
    return this[1];
  }
  set y(newValue: number) {
    this[1] = newValue;
  }
  mapToVector(...args: Array<any>): Vector {
    return new Vector(...this.map(...args));
  }
  multiply(other: Vector) {
    return this.mapToVector((value, index) => value * other[index]);
  }
  plusScalar(scalar: number) {
    return this.mapToVector(value => value + scalar);
  }
  minusScalar(scalar: number) {
    return this.mapToVector(value => value - scalar);
  }
  reverseVector() {
    return new Vector(...super.reverse());
  }
  plus(other: Vector) {
    if (Array.isArray(other)) other = new Vector(...other);
    const biggestVector = this.length > other.length ? this : other;
    const smallestVector = biggestVector === other ? this : other;
    return biggestVector.mapToVector(
      (value, index) =>
        value + (index < smallestVector.length ? smallestVector[index] : 0)
    );
  }
  minus(other: Vector) {
    return this.plus(other.scale(-1));
  }
  scale(scalar: number) {
    return this.mapToVector(value => value * scalar);
  }
  normalize() {
    const length = Math.sqrt(
      this.reduce((lengthSquared, value) => {
        return lengthSquared + value * value;
      }, 0)
    );
    return this.mapToVector(value => value / length);
  }
  inequal(other: Vector) {
    return this.some((value, index) => value !== other[index]);
  }
  equal(other: Vector) {
    return !this.inequal(other);
  }
  lesser(other: Vector) {
    return !this.some((value, index) => value >= other[index]);
  }
  lesserOrEqual(other: Vector) {
    if (this.equal(other)) return true;
    return this.lesser(other);
  }
  greater(other: Vector) {
    return !this.lesserOrEqual(other);
  }

  greaterOrEqual(other: Vector) {
    if (this.equal(other)) return true;
    return this.greater(other);
  }
}
