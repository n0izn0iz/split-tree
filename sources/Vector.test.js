import Vector from "./Vector";

test("constructor throws without arguments", () => {
  expect(() => {
    new Vector();
  }).toThrow();
});

test("constructor matches snapshots with various numbers", () => {
  expect(new Vector(42)).toMatchSnapshot();
  expect(new Vector(1, 2)).toMatchSnapshot();
  expect(new Vector(78, 45, 12)).toMatchSnapshot();
});

test("plusScalar adds the scalar to all components", () => {
  expect(new Vector(42, -42, 21, -21).plusScalar(100.42)).toMatchSnapshot();
});
