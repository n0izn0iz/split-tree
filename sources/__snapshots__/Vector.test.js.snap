// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`constructor matches snapshots with various numbers 1`] = `
Vector [
  42,
]
`;

exports[`constructor matches snapshots with various numbers 2`] = `
Vector [
  1,
  2,
]
`;

exports[`constructor matches snapshots with various numbers 3`] = `
Vector [
  78,
  45,
  12,
]
`;

exports[`plusScalar adds the scalar to all components 1`] = `
Vector [
  142.42000000000002,
  58.42,
  121.42,
  79.42,
]
`;
